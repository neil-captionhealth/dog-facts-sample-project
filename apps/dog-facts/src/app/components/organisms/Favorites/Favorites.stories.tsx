import type { ComponentStory, ComponentMeta } from '@storybook/react';
import { Favorites } from './Favorites';

import { favorites } from './__tests__/mocks';

const Story: ComponentMeta<typeof Favorites> = {
  component: Favorites,
  title: 'Favorites',
  argTypes: {
    setFavorite: {
      action: 'this favorite will be removed',
    },
  },
};
export default Story;

const Template: ComponentStory<typeof Favorites> = (args) => (
  <Favorites {...args} />
);

export const Empty = Template.bind({});
Empty.args = {
  favorites: {},
};

export const Filled = Template.bind({});
Filled.args = {
  favorites,
};
