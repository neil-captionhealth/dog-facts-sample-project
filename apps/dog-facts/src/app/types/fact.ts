export interface IFactFavorites {
  [id: number | string]: {
    description: string;
    isFavorite: boolean;
    fromPage?: number;
  };
}
export interface IFact {
  id: number;
  fact: string;
  length: number;
}

export interface IGetFacts {
  data: IFact[];
  current_page: number;
  first_page_url: string;
  from: number;
  last_page: number;
  last_page_url: string;
  next_page_url: string | null;
  path: string;
  per_page: number;
  prev_page_url: string | null;
  to: number;
  total: number;
}
