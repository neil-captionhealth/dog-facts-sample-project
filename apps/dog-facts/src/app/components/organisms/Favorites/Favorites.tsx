import { SetStateAction, memo } from 'react';

import styled from '@emotion/styled';

import FactBox from '../../molecules/FactBox';

import { IFactFavorites } from '../../../types/fact';

interface Props {
  favorites: IFactFavorites;
  setFavorite: (data: SetStateAction<IFactFavorites>) => void;
}

const Box = styled.div`
  max-width: 390px;
  margin-top: 48px;
  display: flex;
  align-items: center;
  flex-direction: column;
`;

export const Favorites = memo(({ favorites, setFavorite }: Props) => {
  const areFavoritesEmpty = Object.keys(favorites).every(
    (key) => !favorites[key].isFavorite
  );

  return (
    <Box>
      <h1>Favorites</h1>
      {areFavoritesEmpty && (
        <p data-testid="favoritesEmptyLabel">
          You don't have any favorites yet!{' '}
          <span role="img" aria-label="emoji">
            &#129325;
          </span>
        </p>
      )}
      {Object.keys(favorites).map((id) => {
        const description = favorites[id].description;
        const currentFactAdapter = {
          id: Number(id),
          fact: description,
          length: description.length,
        };

        if (!favorites[id].isFavorite) {
          return null;
        }

        return (
          <FactBox
            key={id}
            fact={currentFactAdapter}
            isFavorite
            setFavorite={setFavorite}
          />
        );
      })}
    </Box>
  );
});
