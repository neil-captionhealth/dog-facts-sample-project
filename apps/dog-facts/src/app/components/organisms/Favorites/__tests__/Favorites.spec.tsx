import { render, screen } from '@testing-library/react';
import { SetStateAction } from 'react';

import { IFactFavorites } from '../../../../types/fact';

import '@testing-library/jest-dom';

import { favorites } from './mocks';

import Favorites from '../index';

describe('Favorites', () => {
  const setFavorite: (data: SetStateAction<IFactFavorites>) => void = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render successfully', () => {
    const { baseElement, asFragment } = render(
      <Favorites favorites={favorites} setFavorite={setFavorite} />
    );
    expect(baseElement).toBeTruthy();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render empty label in case no favorites added', () => {
    render(<Favorites favorites={{}} setFavorite={setFavorite} />);

    expect(screen.getByTestId('favoritesEmptyLabel')).toBeInTheDocument();
  });

  it('should call setFavorite to unfavorite element', () => {
    render(<Favorites favorites={favorites} setFavorite={setFavorite} />);

    screen.getAllByTestId('handleFavoriteButton')[0].click();

    expect(setFavorite).toBeCalledTimes(1);
    expect(setFavorite).toBeCalledWith(expect.any(Function));
  });

  it('should render a filled icon', () => {
    render(<Favorites favorites={favorites} setFavorite={setFavorite} />);

    const icon: Partial<HTMLImageElement> =
      screen.getAllByTestId('favoriteIcon')[0];

    expect(icon.src).toMatch(/filled-star.svg/);
  });
});
