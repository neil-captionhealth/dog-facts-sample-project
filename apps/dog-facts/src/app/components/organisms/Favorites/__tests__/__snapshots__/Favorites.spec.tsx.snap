// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites should render successfully 1`] = `
<DocumentFragment>
  <div
    class="css-ruhqkh-Box e1uxv1rf0"
  >
    <h1>
      Favorites
    </h1>
    <div
      class="css-wfh3qg-Box ergcqvm0"
    >
      <span>
        All dogs can be traced back 40 million years ago to a weasel-like animal called the Miacis which dwelled in trees and dens. The Miacis later evolved into the Tomarctus, a direct forbear of the genus Canis, which includes the wolf and jackal as well as the dog.
      </span>
      <button
        class="css-1m826vc-Button e165mwdw1"
        data-testid="handleFavoriteButton"
      >
        <img
          alt="star icon"
          class="css-ef7ogu-Icon e165mwdw0"
          data-testid="favoriteIcon"
          src="filled-star.svg"
        />
        <span>
          Remove from Favorites
        </span>
      </button>
    </div>
    <div
      class="css-wfh3qg-Box ergcqvm0"
    >
      <span>
        In 2003, Dr. Roger Mugford invented the "wagometer", a device that claims to interpret a dog's exact mood by measuring the wag of its tail.
      </span>
      <button
        class="css-1m826vc-Button e165mwdw1"
        data-testid="handleFavoriteButton"
      >
        <img
          alt="star icon"
          class="css-ef7ogu-Icon e165mwdw0"
          data-testid="favoriteIcon"
          src="filled-star.svg"
        />
        <span>
          Remove from Favorites
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
